File: src/tests/unit/agent/canvas/a2ui_contract_fixtures.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

FixtureCase = dict[str, Any]

_FIXTURE_PATH = (
    Path(__file__).resolve().parents[5] / "shared" / "fixtures" / "a2ui-contract-fixtures.json"
)


@lru_cache(maxsize=1)
def load_a2ui_contract_fixture_corpus() -> dict[str, Any]:
    return json.loads(_FIXTURE_PATH.read_text(encoding="utf-8"))


def get_a2ui_contract_case(case_id: str) -> FixtureCase:
    for case in load_a2ui_contract_fixture_corpus()["cases"]:
        if case["id"] == case_id:
            return case
    msg = f"Unknown A2UI contract fixture case: {case_id}"
    raise KeyError(msg)


def iter_backend_contract_cases(*, tier: str | None = None) -> list[FixtureCase]:
    cases: list[FixtureCase] = []
    for case in load_a2ui_contract_fixture_corpus()["cases"]:
        if "backend" not in case.get("targets", []):
            continue
        if tier is not None and case.get("tier") != tier:
            continue
        if not isinstance(case.get("records"), list):
            continue
        cases.append(case)
    return cases


def contract_case_jsonl(case: FixtureCase) -> str:
    records = case.get("records")
    if not isinstance(records, list):
        msg = f"A2UI contract case {case.get('id')} does not define records."
        raise ValueError(msg)
    return "\n".join(json.dumps(record) for record in records)

File: web/src/test/fixtures/a2uiContractFixtures.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import type { A2UIMessageStreamSnapshot } from '@/stores/agent/a2uiMessages';

type ContractTier = 'tier1' | 'tier2' | 'tier3';
type ContractTarget = 'backend' | 'frontend' | 'prompt';

interface ContractIdentity {
  surfaceId?: string;
  metadataSurfaceId?: string;
  hitlRequestId?: string;
}

interface ContractCase {
  id: string;
  tier: ContractTier;
  targets: ContractTarget[];
  description: string;
  records?: Record<string, unknown>[];
  snapshot?: A2UIMessageStreamSnapshot;
  canonicalizes_to: string | null;
  shouldRender: boolean;
  shouldReject: boolean;
  identity?: ContractIdentity;
  historicalSource?: string;
}

interface ContractFixtures {
  contractVersion: number;
  promptGuidance: {
    supportedComponents: string[];
    componentAliases: Record<string, string>;
    rendererOnlyCompatibility: Record<string, string>;
  };
  cases: ContractCase[];
}

const FIXTURE_PATH = resolve(process.cwd(), '../shared/fixtures/a2ui-contract-fixtures.json');

const contractFixtures = JSON.parse(readFileSync(FIXTURE_PATH, 'utf8')) as ContractFixtures;

export function getA2UIContractFixtures(): ContractFixtures {
  return contractFixtures;
}

export function getA2UIContractCase(caseId: string): ContractCase {
  const contractCase = contractFixtures.cases.find((fixtureCase) => fixtureCase.id === caseId);
  if (!contractCase) {
    throw new Error(`Unknown A2UI contract fixture: ${caseId}`);
  }
  return contractCase;
}

export function getA2UIContractMessages(caseId: string): string {
  const contractCase = getA2UIContractCase(caseId);
  if (!contractCase.records) {
    throw new Error(`A2UI contract fixture ${caseId} does not define message records`);
  }
  return contractCase.records.map((record) => JSON.stringify(record)).join('\n');
}

export function getA2UIContractSnapshot(caseId: string): A2UIMessageStreamSnapshot {
  const contractCase = getA2UIContractCase(caseId);
  if (!contractCase.snapshot) {
    throw new Error(`A2UI contract fixture ${caseId} does not define a snapshot`);
  }
  return structuredClone(contractCase.snapshot);
}
